File: platform/android/lynx_android/src/main/java/com/lynx/tasm/gesture/handler/DefaultGestureHandler.java
// Copyright 2023 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.lynx.tasm.gesture.handler;

import android.view.MotionEvent;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.lynx.react.bridge.ReadableMap;
import com.lynx.tasm.behavior.LynxContext;
import com.lynx.tasm.event.LynxTouchEvent;
import com.lynx.tasm.gesture.GestureArenaMember;
import com.lynx.tasm.gesture.common.GestureExtraBundle;
import com.lynx.tasm.gesture.detector.GestureDetector;
import java.util.HashMap;

/**
 * The DefaultGestureHandler class is a concrete implementation of the BaseGestureHandler class
 * for handling default gestures. It handle default pan and fling gesture for scrolling component,
 * which provides aligned scroll rate and decay curves on android and iOS.
 */
public class DefaultGestureHandler extends BaseGestureHandler {
  private final HashMap<String, Object> mEventParams;

  // record x axis when last action
  private float mLastX = 0f;
  // record x axis when last action
  private float mLastY = 0f;
  // is onBegin invoked or not
  private boolean mIsInvokedBegin = false;
  // is onStart invoked or not
  private boolean mIsInvokedStart = false;
  // is onEnd invoked or not
  private boolean mIsInvokedEnd = false;
  // default tap slop
  private int mTapSlop = 3;

  // record last lynx touch event
  private LynxTouchEvent mLastTouchEvent;

  /**
   * Constructs a DefaultGestureHandler object with the specified properties.
   *
   * @param sign                the sign indicating the direction of the gesture
   * @param lynxContext         the LynxContext associated with the gesture handler
   * @param gestureDetector     the GestureDetector associated with the gesture handler
   * @param gestureArenaMember  the GestureArenaMember associated with the gesture handler
   */
  public DefaultGestureHandler(int sign, LynxContext lynxContext,
      @NonNull GestureDetector gestureDetector, GestureArenaMember gestureArenaMember) {
    super(sign, lynxContext, gestureDetector, gestureArenaMember);
    handleConfigMap(gestureDetector.getConfigMap());
    mEventParams = new HashMap<>();
  }

  @Override
  protected void handleConfigMap(@Nullable ReadableMap config) {
    if (config == null) {
      return;
    }
    mTapSlop = config.getInt(GestureConstants.TAP_SLOP, 3);
  }

  @Override
  protected void onHandle(@Nullable MotionEvent event, @Nullable LynxTouchEvent lynxTouchEvent,
      float flingDeltaX, float flingDeltaY, boolean handleBySimultaneous,
      @Nullable GestureExtraBundle extraBundle) {
    mLastTouchEvent = lynxTouchEvent;

    if (handleBySimultaneous && extraBundle != null) {
      if (extraBundle.isConsumedGesture()) {
        if (mGestureArenaMember != null) {
          mGestureArenaMember.onGestureScrollBy(
              extraBundle.getSimultaneousDeltaX(), extraBundle.getSimultaneousDeltaY());
          mGestureArenaMember.onInvalidate();
        }
      }
      return;
    }

    if (mStatus >= GestureConstants.LYNX_STATE_FAIL) {
      onEnd(mLastX, mLastY, mLastTouchEvent);
      return;
    }
    if (event != null) {
      // in pan status
      switch (event.getActionMasked()) {
        case MotionEvent.ACTION_DOWN:
          mLastX = event.getX();
          mLastY = event.getY();
          mIsInvokedEnd = false;
          begin();
          onBegin(mLastX, mLastY, lynxTouchEvent);
          break;
        case MotionEvent.ACTION_MOVE:
          float deltaX = mLastX - event.getX();
          float deltaY = mLastY - event.getY();

          if (deltaX == 0 && deltaY == 0) {
            break;
          }
          if (extraBundle != null
              && extraBundle.getGestureDirection() == GestureConstants.DIRECTION_UNDETERMINED) {
            extraBundle.setGestureDirection(Math.abs(deltaX) > Math.abs(deltaY)
                    ? GestureConstants.DIRECTION_HORIZONTAL
                    : GestureConstants.DIRECTION_VERTICAL);
          }
          if (extraBundle != null
              && extraBundle.getGestureDirection() != GestureConstants.DIRECTION_UNDETERMINED) {
            if (extraBundle.getGestureDirection() == GestureConstants.DIRECTION_HORIZONTAL) {
              deltaY = 0;
            } else {
              deltaX = 0;
            }
          }

          if (mStatus == GestureConstants.LYNX_STATE_INIT) {
            onBegin(mLastX, mLastY, lynxTouchEvent);
            if (mStatus <= GestureConstants.LYNX_STATE_BEGIN) {
              activate();
            }
          } else {
            if (shouldFail(deltaX, deltaY, extraBundle)) {
              // consume last delta to arrive start or end
              if (mGestureArenaMember != null && !mGestureArenaMember.isAtBorder(true)
                  && !mGestureArenaMember.isAtBorder(false)) {
                onUpdate(deltaX, deltaY, lynxTouchEvent, extraBundle);
              }
              fail();
              onEnd(mLastX, mLastY, lynxTouchEvent);
            } else {
              activate();
              onUpdate(deltaX, deltaY, lynxTouchEvent, extraBundle);
            }
          }
          mLastX = event.getX();
          mLastY = event.getY();
          break;
        case MotionEvent.ACTION_UP:
          if (mStatus == GestureConstants.LYNX_STATE_ACTIVE && flingDeltaX == Float.MIN_VALUE
              && flingDeltaY == Float.MIN_VALUE) {
            fail();
            onEnd(0, 0, null);
          }
          break;
        default:
          break;
      }
    } else {
      if (mGestureArenaMember != null) {
        mGestureArenaMember.onInvalidate();
      }
      // in fling status
      if (mStatus == GestureConstants.LYNX_STATE_ACTIVE && flingDeltaX == Float.MIN_VALUE
          && flingDeltaY == Float.MIN_VALUE) {
        fail();
        onEnd(0, 0, null);
        return;
      }

      if (extraBundle != null
          && extraBundle.getGestureDirection() != GestureConstants.DIRECTION_UNDETERMINED) {
        if (extraBundle.getGestureDirection() == GestureConstants.DIRECTION_HORIZONTAL) {
          flingDeltaY = 0;
        } else {
          flingDeltaX = 0;
        }
      }

      if (shouldFail(flingDeltaX, flingDeltaY, extraBundle)) {
        // consume last delta to arrive start or end
        if (mGestureArenaMember != null && !mGestureArenaMember.isAtBorder(true)
            && !mGestureArenaMember.isAtBorder(false)) {
          onUpdate(flingDeltaX, flingDeltaY, null, extraBundle);
        }
        fail();
        onEnd(flingDeltaX, flingDeltaY, null);
      } else {
        if (mStatus == GestureConstants.LYNX_STATE_INIT) {
          onBegin(mLastX, mLastY, lynxTouchEvent);
          if (mStatus <= GestureConstants.LYNX_STATE_BEGIN) {
            activate();
          }
        }
        onUpdate(flingDeltaX, flingDeltaY, null, extraBundle);
      }
      if (mGestureArenaMember != null) {
        mGestureArenaMember.onInvalidate();
      }
    }
  }

  private boolean shouldFail(float deltaX, float deltaY, @Nullable GestureExtraBundle extraBundle) {
    if (extraBundle != null) {
      if (extraBundle.getGestureDirection() == GestureConstants.DIRECTION_HORIZONTAL
          && mGestureArenaMember.getScrollContainerDirection()
              != GestureConstants.DIRECTION_HORIZONTAL) {
        return true;
      }
      if (extraBundle.getGestureDirection() == GestureConstants.DIRECTION_VERTICAL
          && mGestureArenaMember.getScrollContainerDirection()
              != GestureConstants.DIRECTION_VERTICAL) {
        return true;
      }
    }
    return !mGestureArenaMember.canConsumeGesture(deltaX, deltaY);
  }

  protected HashMap<String, Object> getEventParamsInActive(float deltaX, float deltaY) {
    // Set the event parameters for the active state
    mEventParams.put("scrollX", px2dip(mGestureArenaMember.getMemberScrollX()));
    mEventParams.put("scrollY", px2dip(mGestureArenaMember.getMemberScrollY()));
    mEventParams.put("deltaX", px2dip(deltaX));
    mEventParams.put("deltaY", px2dip(deltaY));
    mEventParams.put("isAtStart", mGestureArenaMember.isAtBorder(true));
    mEventParams.put("isAtEnd", mGestureArenaMember.isAtBorder(false));
    return mEventParams;
  }

  @Override
  public void fail() {
    super.fail();
    if (mLastTouchEvent == null) {
      onEnd(0, 0, null);
    } else {
      onEnd(mLastX, mLastY, mLastTouchEvent);
    }
  }

  @Override
  public void activate() {
    super.activate();
  }

  @Override
  public void end() {
    super.end();
    if (mLastTouchEvent == null) {
      onEnd(0, 0, null);
    } else {
      onEnd(mLastX, mLastY, mLastTouchEvent);
    }
  }

  @Override
  public void reset() {
    super.reset();
    mLastX = 0.0f;
    mLastY = 0.0f;
    mIsInvokedBegin = false;
    mIsInvokedStart = false;
    mIsInvokedEnd = false;
  }

  @Override
  protected void onBegin(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnBeginEnable() || mIsInvokedBegin) {
      return;
    }
    mIsInvokedBegin = true;
    sendGestureEvent(GestureConstants.ON_BEGIN, getEventParamsInActive(0, 0));
  }

  @Override
  protected void onUpdate(float deltaX, float deltaY, @Nullable LynxTouchEvent event,
      @Nullable GestureExtraBundle extraBundle) {
    if (mGestureArenaMember != null) {
      mGestureArenaMember.onGestureScrollBy(deltaX, deltaY);
    }

    if (extraBundle != null) {
      extraBundle.setConsumedGesture(true);
      extraBundle.setSimultaneousDeltaX(deltaX);
      extraBundle.setSimultaneousDeltaY(deltaY);
    }

    if (Math.abs((int) deltaX) > mTapSlop || Math.abs((int) deltaY) > mTapSlop) {
      mLynxContext.onGestureRecognized(mSign);
    }

    if (!isOnUpdateEnable()) {
      return;
    }
    sendGestureEvent(GestureConstants.ON_UPDATE, getEventParamsInActive(deltaX, deltaY));
  }

  @Override
  protected void onStart(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnStartEnable() || mIsInvokedStart || !mIsInvokedBegin) {
      return;
    }
    mIsInvokedStart = true;
    sendGestureEvent(GestureConstants.ON_START, getEventParamsInActive(x, y));
  }

  @Override
  protected void onEnd(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnEndEnable() || mIsInvokedEnd || !mIsInvokedBegin) {
      return;
    }
    mIsInvokedEnd = true;
    sendGestureEvent(GestureConstants.ON_END, getEventParamsInActive(0, 0));
  }
}

File: platform/harmony/lynx_harmony/src/main/cpp/gesture/handler/default_gesture_handler.cc
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "platform/harmony/lynx_harmony/src/main/cpp/gesture/handler/default_gesture_handler.h"

#include <limits>

#include "platform/harmony/lynx_harmony/src/main/cpp/event/gesture_event.h"
#include "platform/harmony/lynx_harmony/src/main/cpp/event/touch_event.h"
#include "platform/harmony/lynx_harmony/src/main/cpp/gesture/gesture_arena_member.h"
#include "platform/harmony/lynx_harmony/src/main/cpp/lynx_context.h"

namespace lynx {
namespace tasm {
namespace harmony {
DefaultGestureHandler::DefaultGestureHandler(
    int sign, LynxContext* lynx_context,
    std::shared_ptr<GestureDetector> gesture_detector,
    std::weak_ptr<GestureArenaMember> gesture_arena_member)
    : BaseGestureHandler(sign, lynx_context, gesture_detector,
                         gesture_arena_member),

      is_invoked_begin_(false),
      is_invoked_start_(false),
      is_invoked_end_(false) {
  HandleConfigMap(gesture_detector->gesture_config());
}

void DefaultGestureHandler::HandleConfigMap(const lepus::Value& config) {}

void DefaultGestureHandler::OnHandle(
    const ArkUI_UIInputEvent* event,
    std::shared_ptr<TouchEvent> lynx_touch_event, float fling_delta_x,
    float fling_delta_y) {
  last_touch_event_ = lynx_touch_event;
  if (status_ >= GestureConstants::LYNX_STATE_FAIL) {
    OnEnd(last_x_, last_y_, last_touch_event_);
    return;
  }

  if (event != nullptr) {
    int32_t type = OH_ArkUI_UIInputEvent_GetAction(event);
    if (type == UI_TOUCH_EVENT_ACTION_DOWN) {
      last_x_ = OH_ArkUI_PointerEvent_GetXByIndex(event, 0);
      last_y_ = OH_ArkUI_PointerEvent_GetYByIndex(event, 0);
      is_invoked_end_ = false;
      Begin();
      OnBegin(last_x_, last_y_, lynx_touch_event);
    } else if (type == UI_TOUCH_EVENT_ACTION_MOVE) {
      float delta_x = last_x_ - OH_ArkUI_PointerEvent_GetXByIndex(event, 0);
      float delta_y = last_y_ - OH_ArkUI_PointerEvent_GetYByIndex(event, 0);
      if (status_ == GestureConstants::LYNX_STATE_INIT) {
        OnBegin(last_x_, last_y_, lynx_touch_event);
        Activate();
      } else {
        if (ShouldFail(delta_x, delta_y)) {
          OnUpdate(delta_x, delta_y, lynx_touch_event);
          Fail();
          OnEnd(last_x_, last_y_, lynx_touch_event);
        } else {
          Activate();
          OnUpdate(delta_x, delta_y, lynx_touch_event);
        }
      }
      last_x_ = OH_ArkUI_PointerEvent_GetXByIndex(event, 0);
      last_y_ = OH_ArkUI_PointerEvent_GetYByIndex(event, 0);
    } else if (type == UI_TOUCH_EVENT_ACTION_UP) {
      if (status_ == GestureConstants::LYNX_STATE_ACTIVE &&
          fling_delta_x == std::numeric_limits<float>::min() &&
          fling_delta_y == std::numeric_limits<float>::min()) {
        Fail();
        OnEnd(0, 0, nullptr);
      }
    }
  } else {
    if (status_ == GestureConstants::LYNX_STATE_ACTIVE &&
        fling_delta_x == std::numeric_limits<float>::min() &&
        fling_delta_y == std::numeric_limits<float>::min()) {
      Fail();
      OnEnd(0, 0, nullptr);
      return;
    }
    if (ShouldFail(fling_delta_x, fling_delta_y)) {
      OnUpdate(fling_delta_x, fling_delta_y, nullptr);
      Fail();
      OnEnd(fling_delta_x, fling_delta_y, nullptr);
    } else {
      if (status_ == GestureConstants::LYNX_STATE_INIT) {
        OnBegin(last_x_, last_y_, lynx_touch_event);
        Activate();
        return;
      }
      OnUpdate(fling_delta_x, fling_delta_y, nullptr);
    }
  }
}

bool DefaultGestureHandler::ShouldFail(float delta_x, float delta_y) {
  auto member = gesture_arena_member_.lock();
  if (!member) {
    return true;
  }
  return !member->CanConsumeGesture(delta_x, delta_y);
}

lepus::Value DefaultGestureHandler::GetEventParamsInActive(
    const std::shared_ptr<TouchEvent>& touch_event, float delta_x,
    float delta_y) {
  auto event_params_ = lepus::Dictionary::Create();

  event_params_->SetValue("scrollX", gesture_arena_member_.lock()->ScrollX());
  event_params_->SetValue("scrollY", gesture_arena_member_.lock()->ScrollY());
  event_params_->SetValue("deltaX", delta_x);
  event_params_->SetValue("deltaY", delta_y);
  event_params_->SetValue("isAtStart",
                          gesture_arena_member_.lock()->IsAtBorder(true));
  event_params_->SetValue("isAtEnd",
                          gesture_arena_member_.lock()->IsAtBorder(false));
  auto additional_params = GetEventParamsFromTouchEvent(touch_event);
  for (auto& param : *additional_params.Table()) {
    event_params_->SetValue(param.first, std::move(param.second));
  }
  return lepus::Value(std::move(event_params_));
}

void DefaultGestureHandler::Fail() {
  if (status_ != GestureConstants::LYNX_STATE_FAIL) {
    status_ = GestureConstants::LYNX_STATE_FAIL;
  }
  if (last_touch_event_ == nullptr) {
    OnEnd(0, 0, nullptr);
  } else {
    OnEnd(last_x_, last_y_, last_touch_event_);
  }
}

void DefaultGestureHandler::End() {
  if (status_ != GestureConstants::LYNX_STATE_END) {
    status_ = GestureConstants::LYNX_STATE_END;
  }
  if (last_touch_event_ == nullptr) {
    OnEnd(0, 0, nullptr);
  } else {
    OnEnd(last_x_, last_y_, last_touch_event_);
  }
}

void DefaultGestureHandler::Reset() {
  BaseGestureHandler::Reset();
  last_x_ = 0;
  last_y_ = 0;
  is_invoked_begin_ = false;
  is_invoked_start_ = false;
  is_invoked_end_ = false;
}

void DefaultGestureHandler::OnBegin(float x, float y,
                                    std::shared_ptr<TouchEvent> event) {
  if (!IsOnBeginEnable() || is_invoked_begin_) {
    return;
  }
  is_invoked_begin_ = true;
  SendGestureEvent(GestureConstants::ON_BEGIN,
                   GetEventParamsInActive(event, 0, 0));
}

void DefaultGestureHandler::OnUpdate(float delta_x, float delta_y,
                                     std::shared_ptr<TouchEvent> event) {
  auto member = gesture_arena_member_.lock();
  if (member != nullptr) {
    member->ScrollBy(delta_x, delta_y);
  }
  if (!IsOnUpdateEnable()) {
    return;
  }
  SendGestureEvent(GestureConstants::ON_UPDATE,
                   GetEventParamsInActive(event, delta_x, delta_y));
}

void DefaultGestureHandler::OnStart(float x, float y,
                                    std::shared_ptr<TouchEvent> event) {
  if (!IsOnStartEnable() || is_invoked_start_ || !is_invoked_begin_) {
    return;
  }
  is_invoked_start_ = true;
  SendGestureEvent(GestureConstants::ON_START,
                   GetEventParamsInActive(event, x, y));
}

void DefaultGestureHandler::OnEnd(float x, float y,
                                  std::shared_ptr<TouchEvent> event) {
  if (!IsOnEndEnable() || is_invoked_end_ || !is_invoked_begin_) {
    return;
  }
  is_invoked_end_ = true;
  SendGestureEvent(GestureConstants::ON_END,
                   GetEventParamsInActive(event, 0, 0));
}

}  // namespace harmony
}  // namespace tasm
}  // namespace lynx
